File: principal-engineer/csharp/tests/OrderService.Tests/OrderApiTests.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc.Testing;
using Xunit;

namespace OrderService.Tests;

/// <summary>
/// Behavioural tests for the Order Processing Service. They specify the intended
/// behaviour around idempotency, inventory reservation/oversell, cancellation
/// compensation and order totals. Several fail against the starting code.
/// Tests tagged [Trait("Category","Bonus")] cover the bonus fulfillment saga.
/// </summary>
public class OrderApiTests : IClassFixture<WebApplicationFactory<Program>>
{
    private readonly WebApplicationFactory<Program> _factory;
    public OrderApiTests(WebApplicationFactory<Program> factory) => _factory = factory;

    private HttpClient NewClient() => _factory.WithWebHostBuilder(_ => { }).CreateClient();
    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    private record LineDto(string sku, int quantity, decimal unitPrice);
    private record OrderResponse(string id, string status, decimal total, List<LineDto>? lines);
    private record InventoryResponse(string sku, int available, int reserved);

    private static HttpRequestMessage PostOrder(object body, string? idempotencyKey = null)
    {
        var req = new HttpRequestMessage(HttpMethod.Post, "/orders")
        {
            Content = JsonContent.Create(body)
        };
        if (idempotencyKey is not null) req.Headers.Add("Idempotency-Key", idempotencyKey);
        return req;
    }

    private static object Order(params object[] lines) => new { lines };
    private static object Line(string sku, int qty, decimal price) => new { sku, quantity = qty, unitPrice = price };

    [Fact]
    public async Task Health_Returns_Ok()
    {
        var res = await NewClient().GetAsync("/health");
        Assert.Equal(HttpStatusCode.OK, res.StatusCode);
    }

    [Fact] // regression — should already pass
    public async Task Create_Order_Returns_201_Placed_With_Id()
    {
        var client = NewClient();
        var res = await client.SendAsync(PostOrder(Order(Line("SKU-1", 1, 10m))));
        Assert.Equal(HttpStatusCode.Created, res.StatusCode);
        var order = await res.Content.ReadFromJsonAsync<OrderResponse>(Json);
        Assert.NotNull(order);
        Assert.False(string.IsNullOrWhiteSpace(order!.id));
        Assert.Equal("Placed", order.status);
    }

    [Fact] // Defect #4 — total ignores quantity
    public async Task Order_Total_Uses_Quantity()
    {
        var client = NewClient();
        var res = await client.SendAsync(PostOrder(Order(Line("SKU-1", 2, 10m), Line("SKU-2", 3, 5m))));
        res.EnsureSuccessStatusCode();
        var order = await res.Content.ReadFromJsonAsync<OrderResponse>(Json);
        Assert.Equal(35m, order!.total); // 2*10 + 3*5
    }

    [Fact] // Defect #2 — reservation not applied
    public async Task Valid_Order_Reserves_Inventory()
    {
        var client = NewClient();
        var res = await client.SendAsync(PostOrder(Order(Line("SKU-1", 3, 10m))));
        res.EnsureSuccessStatusCode();

        var inv = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(7, inv!.available);
        Assert.Equal(3, inv.reserved);
    }

    [Fact] // Defect #2 — oversell must be rejected
    public async Task Oversell_Is_Rejected_And_Inventory_Unchanged()
    {
        var client = NewClient();
        var res = await client.SendAsync(PostOrder(Order(Line("SKU-1", 99, 10m))));
        Assert.Equal(HttpStatusCode.Conflict, res.StatusCode);

        var inv = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(10, inv!.available);
        Assert.Equal(0, inv.reserved);
    }

    [Fact] // Defect #2 — unknown sku rejected
    public async Task Unknown_Sku_Is_Rejected()
    {
        var client = NewClient();
        var res = await client.SendAsync(PostOrder(Order(Line("SKU-X", 1, 10m))));
        Assert.Equal(HttpStatusCode.Conflict, res.StatusCode);
    }

    [Fact] // Defect #1 — idempotency
    public async Task Repeated_IdempotencyKey_Returns_Same_Order_And_Reserves_Once()
    {
        var client = NewClient();
        var first = await client.SendAsync(PostOrder(Order(Line("SKU-1", 2, 10m)), "key-1"));
        var second = await client.SendAsync(PostOrder(Order(Line("SKU-1", 2, 10m)), "key-1"));

        var o1 = await first.Content.ReadFromJsonAsync<OrderResponse>(Json);
        var o2 = await second.Content.ReadFromJsonAsync<OrderResponse>(Json);
        Assert.Equal(o1!.id, o2!.id);

        var inv = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(8, inv!.available); // reserved exactly once
    }

    [Fact]
    public async Task Get_Missing_Order_Returns_404()
    {
        var res = await NewClient().GetAsync("/orders/nope");
        Assert.Equal(HttpStatusCode.NotFound, res.StatusCode);
    }

    [Fact] // Defect #3 — cancel restores inventory
    public async Task Cancel_Restores_Inventory()
    {
        var client = NewClient();
        var place = await client.SendAsync(PostOrder(Order(Line("SKU-1", 4, 10m))));
        var order = await place.Content.ReadFromJsonAsync<OrderResponse>(Json);

        // Placing must reserve stock first (10 - 4 = 6).
        var afterPlace = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(6, afterPlace!.available);

        var cancel = await client.PostAsync($"/orders/{order!.id}/cancel", null);
        Assert.Equal(HttpStatusCode.OK, cancel.StatusCode);
        var cancelled = await cancel.Content.ReadFromJsonAsync<OrderResponse>(Json);
        Assert.Equal("Cancelled", cancelled!.status);

        var inv = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(10, inv!.available);
        Assert.Equal(0, inv.reserved);
    }

    [Fact] // Defect #3 — cancel is idempotent (no double restore)
    public async Task Cancel_Twice_Does_Not_Double_Restore()
    {
        var client = NewClient();
        var place = await client.SendAsync(PostOrder(Order(Line("SKU-1", 4, 10m))));
        var order = await place.Content.ReadFromJsonAsync<OrderResponse>(Json);

        // Placing must reserve stock first (10 - 4 = 6).
        var afterPlace = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(6, afterPlace!.available);

        await client.PostAsync($"/orders/{order!.id}/cancel", null);
        await client.PostAsync($"/orders/{order.id}/cancel", null);

        var inv = await client.GetFromJsonAsync<InventoryResponse>("/inventory/SKU-1", Json);
        Assert.Equal(10, inv!.available); // not 14
    }

    [Fact]
    [Trait("Category", "Bonus")]
    public async Task Bonus_Drain_Events_Fulfills_Order()
    {
        var client = NewClient();
        var place = await client.SendAsync(PostOrder(Order(Line("SKU-1", 1, 10m))));
        var order = await place.Content.ReadFromJsonAsync<OrderResponse>(Json);

        var drain = await client.PostAsync("/internal/drain-events", null);
        Assert.Equal(HttpStatusCode.OK, drain.StatusCode);

        var fetched = await client.GetFromJsonAsync<OrderResponse>($"/orders/{order!.id}", Json);
        Assert.Equal("Fulfilled", fetched!.status);
    }

    [Fact]
    [Trait("Category", "Bonus")]
    public async Task Bonus_Drain_Events_Publishes_OrderFulfilled()
    {
        var client = NewClient();
        var place = await client.SendAsync(PostOrder(Order(Line("SKU-2", 1, 5m))));
        var order = await place.Content.ReadFromJsonAsync<OrderResponse>(Json);

        await client.PostAsync("/internal/drain-events", null);

        var events = await client.GetFromJsonAsync<List<string>>("/internal/events", Json);
        Assert.NotNull(events);
        Assert.Contains(events!, e => e.Contains("OrderFulfilled"));
    }
}

File: principal-engineer/java/src/test/java/com/interview/order/OrderApiTests.java
package com.interview.order;

import com.fasterxml.jackson.databind.JsonNode;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.test.annotation.DirtiesContext;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Behavioural tests for the Order Processing Service. They specify the intended
 * behaviour around idempotency, inventory reservation/oversell, cancellation
 * compensation and order totals. Several fail against the starting code.
 * Tests tagged "Bonus" cover the bonus fulfillment saga.
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
class OrderApiTests {
    @Autowired
    private TestRestTemplate client;

    @Test
    void healthReturnsOk() {
        ResponseEntity<JsonNode> response = client.getForEntity("/health", JsonNode.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
    }

    @Test
    void createOrderReturns201PlacedWithId() {
        ResponseEntity<JsonNode> response = postOrder(order(line("SKU-1", 1, 10.0)));
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        assertThat(response.getBody().get("id").asText()).isNotBlank();
        assertThat(response.getBody().get("status").asText()).isEqualTo("Placed");
    }

    @Test
    void orderTotalUsesQuantity() {
        ResponseEntity<JsonNode> response = postOrder(order(line("SKU-1", 2, 10.0), line("SKU-2", 3, 5.0)));
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);
        assertThat(response.getBody().get("total").asDouble()).isEqualTo(35.0);
    }

    @Test
    void validOrderReservesInventory() {
        ResponseEntity<JsonNode> response = postOrder(order(line("SKU-1", 3, 10.0)));
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        JsonNode inventory = getJson("/inventory/SKU-1");
        assertThat(inventory.get("available").asInt()).isEqualTo(7);
        assertThat(inventory.get("reserved").asInt()).isEqualTo(3);
    }

    @Test
    void oversellIsRejectedAndInventoryUnchanged() {
        ResponseEntity<JsonNode> response = postOrder(order(line("SKU-1", 99, 10.0)));
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CONFLICT);

        JsonNode inventory = getJson("/inventory/SKU-1");
        assertThat(inventory.get("available").asInt()).isEqualTo(10);
        assertThat(inventory.get("reserved").asInt()).isEqualTo(0);
    }

    @Test
    void unknownSkuIsRejected() {
        ResponseEntity<JsonNode> response = postOrder(order(line("SKU-X", 1, 10.0)));
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.CONFLICT);
    }

    @Test
    void repeatedIdempotencyKeyReturnsSameOrderAndReservesOnce() {
        String payload = order(line("SKU-1", 2, 10.0));

        ResponseEntity<JsonNode> first = postOrder(payload, "key-1");
        ResponseEntity<JsonNode> second = postOrder(payload, "key-1");

        assertThat(first.getBody().get("id").asText()).isEqualTo(second.getBody().get("id").asText());

        JsonNode inventory = getJson("/inventory/SKU-1");
        assertThat(inventory.get("available").asInt()).isEqualTo(8);
        assertThat(inventory.get("reserved").asInt()).isEqualTo(2);
    }

    @Test
    void getMissingOrderReturns404() {
        ResponseEntity<JsonNode> response = client.getForEntity("/orders/nope", JsonNode.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    void cancelRestoresInventory() {
        ResponseEntity<JsonNode> place = postOrder(order(line("SKU-1", 4, 10.0)));
        assertThat(place.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        // Placing must reserve stock first (10 - 4 = 6).
        JsonNode afterPlace = getJson("/inventory/SKU-1");
        assertThat(afterPlace.get("available").asInt()).isEqualTo(6);

        ResponseEntity<JsonNode> cancel = client.postForEntity(
                "/orders/" + place.getBody().get("id").asText() + "/cancel", null, JsonNode.class);
        assertThat(cancel.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(cancel.getBody().get("status").asText()).isEqualTo("Cancelled");

        JsonNode inventory = getJson("/inventory/SKU-1");
        assertThat(inventory.get("available").asInt()).isEqualTo(10);
        assertThat(inventory.get("reserved").asInt()).isEqualTo(0);
    }

    @Test
    void cancelTwiceDoesNotDoubleRestore() {
        ResponseEntity<JsonNode> place = postOrder(order(line("SKU-1", 4, 10.0)));
        assertThat(place.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        // Placing must reserve stock first (10 - 4 = 6).
        JsonNode afterPlace = getJson("/inventory/SKU-1");
        assertThat(afterPlace.get("available").asInt()).isEqualTo(6);

        String cancelPath = "/orders/" + place.getBody().get("id").asText() + "/cancel";
        assertThat(client.postForEntity(cancelPath, null, JsonNode.class).getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(client.postForEntity(cancelPath, null, JsonNode.class).getStatusCode()).isEqualTo(HttpStatus.OK);

        JsonNode inventory = getJson("/inventory/SKU-1");
        assertThat(inventory.get("available").asInt()).isEqualTo(10);
        assertThat(inventory.get("reserved").asInt()).isEqualTo(0);
    }

    @Test
    void bonusDrainEventsFulfillsOrder() {
        ResponseEntity<JsonNode> place = postOrder(order(line("SKU-1", 1, 10.0)));
        assertThat(place.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        ResponseEntity<JsonNode> drain = client.postForEntity("/internal/drain-events", null, JsonNode.class);
        assertThat(drain.getStatusCode()).isEqualTo(HttpStatus.OK);

        JsonNode fetched = getJson("/orders/" + place.getBody().get("id").asText());
        assertThat(fetched.get("status").asText()).isEqualTo("Fulfilled");
    }

    @Test
    void bonusDrainEventsPublishesOrderFulfilled() {
        ResponseEntity<JsonNode> place = postOrder(order(line("SKU-2", 1, 5.0)));
        assertThat(place.getStatusCode()).isEqualTo(HttpStatus.CREATED);

        client.postForEntity("/internal/drain-events", null, JsonNode.class);

        ResponseEntity<List<String>> eventsResponse = client.exchange(
                "/internal/events", HttpMethod.GET, null, new ParameterizedTypeReference<>() {
                });
        assertThat(eventsResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(eventsResponse.getBody()).anyMatch(event -> event.contains("OrderFulfilled"));
    }

    private ResponseEntity<JsonNode> postOrder(String body) {
        return postOrder(body, null);
    }

    private ResponseEntity<JsonNode> postOrder(String body, String idempotencyKey) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        if (idempotencyKey != null) {
            headers.add("Idempotency-Key", idempotencyKey);
        }
        return client.postForEntity("/orders", new HttpEntity<>(body, headers), JsonNode.class);
    }

    private JsonNode getJson(String path) {
        ResponseEntity<JsonNode> response = client.getForEntity(path, JsonNode.class);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        return response.getBody();
    }

    private static String order(String... lines) {
        return "{\"lines\":[" + String.join(",", lines) + "]}";
    }

    private static String line(String sku, int quantity, double unitPrice) {
        return "{\"sku\":\"" + sku + "\",\"quantity\":" + quantity + ",\"unitPrice\":" + unitPrice + "}";
    }
}
